File: answers/leetcode/Longest Increasing Path in a Matrix.cpp
//@result 137 / 137 test cases passed. Status: Accepted Runtime: 84 ms Submitted: 0 minutes ago You are here! Your runtime beats 60.93% of cppsubmissions.

class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        if (0 == matrix.size() || 0 == matrix[0].size()) {
            return 0;
        }
        vector<vector<int> > length_mat(matrix.size(), vector<int> (matrix[0].size(), INT_MAX));
        int longest = 0;
        for (size_t i = 0; i < matrix.size(); ++ i) {
            for (size_t j = 0; j < matrix[i].size(); ++ j) {
                dfs(i, j, matrix, length_mat);
                longest = max(longest, length_mat[i][j]);
            }
        }
        return longest;
    }
    void dfs(size_t i, size_t j, const vector<vector<int> > &matrix, vector<vector<int> > &length_mat) {
        if (i >= matrix.size() || j >= matrix[i].size() || length_mat[i][j] < INT_MAX) {
            return;
        }
        int longest = 1;
        if (i > 0 && matrix[i][j] < matrix[i - 1][j]) {
            dfs(i - 1, j, matrix, length_mat);
            longest = max(longest, length_mat[i - 1][j] + 1);
        }
        if (i + 1 < matrix.size() && matrix[i][j] < matrix[i + 1][j]) {
            dfs(i + 1, j, matrix, length_mat);
            longest = max(longest, length_mat[i + 1][j] + 1);
        }
        if (j > 0 && matrix[i][j] < matrix[i][j - 1]) {
            dfs(i, j - 1, matrix, length_mat);
            longest = max(longest, length_mat[i][j - 1] + 1);
        }
        if (j + 1 < matrix[i].size() && matrix[i][j] < matrix[i][j + 1]) {
            dfs(i, j + 1, matrix, length_mat);
            longest = max(longest, length_mat[i][j + 1] + 1);
        }
        length_mat[i][j] = longest;
        //cout << "debug " << i << " " << j << " " << length_mat[i][j] << endl;
    }
};

File: answers/leetcode/Longest Increasing Path in a Matrix/Longest Increasing Path in a Matrix.java
class Solution {
    public int longestIncreasingPath(int[][] matrix) {
        if (matrix == null || matrix.length == 0) {
            return 0;
        }
        int[][] dp = new int[matrix.length][matrix[0].length];
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                this.dfs(1, i, j, dp, matrix);
            }
        }
        int longest = 0;
        for (int i = 0; i < dp.length; i++) {
            for (int j = 0; j < dp[i].length; j++) {
                longest = Math.max(longest, dp[i][j]);
            }
        }
        return longest;
    }

    private void dfs(int distance, int x, int y, int[][] dp, int[][] matrix) {
        if (distance <= dp[x][y]) {
            return;
        }

        dp[x][y] = distance;
        if (x > 0 && matrix[x][y] < matrix[x - 1][y]) {
            this.dfs(distance + 1, x - 1, y, dp, matrix);
        }
        if (x + 1 < matrix.length && matrix[x][y] < matrix[x + 1][y]) {
            this.dfs(distance + 1, x + 1, y, dp, matrix);
        }
        if (y > 0 && matrix[x][y] < matrix[x][y - 1]) {
            this.dfs(distance + 1, x, y - 1, dp, matrix);
        }
        if (y + 1 < matrix[x].length && matrix[x][y] < matrix[x][y + 1]) {
            this.dfs(distance + 1, x, y + 1, dp, matrix);
        }
    }
}